src/pu-main.c

```diff
@@ -5,168 +5,258 @@
 #include "pu-emmc.h"
+#include "pu-error.h"
 #include "pu-flash.h"
 #include "pu-mount.h"
+#include "pu-package.h"
 #include "pu-utils.h"
 #include "pu-version.h"
 
 static gboolean arg_debug = FALSE;
-static gboolean arg_version = FALSE;
-static gboolean arg_skip_checksums = FALSE;
-static gchar *arg_config = NULL;
-static gchar *arg_device = NULL;
-static gchar *arg_prefix = NULL;
+static gboolean arg_install_skip_checksums = FALSE;
+static gchar *arg_package_directory = NULL;
+static gboolean arg_package_force = FALSE;
+static gboolean arg_show_size = FALSE;
+
+static inline gboolean
+error_not_root(GError **error)
+{
+    g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                "%s must be run as root", g_get_prgname());
+    return FALSE;
+}
+
+static inline gboolean
+error_out(const gchar *mountpoint)
+{
+    pu_package_umount(mountpoint, NULL);
+    return FALSE;
+}
 
 static gboolean
-arg_parse_remaining(const gchar *option_name,
-                    const gchar *value,
-                    G_GNUC_UNUSED gpointer data,
-                    GError **error)
+cmd_install(PuCommandContext *context,
+            GError **error)
 {
-    g_debug("%s(option_name=\"%s\" value=\"%s\")", __func__, option_name, value);
+    g_autoptr(PuConfig) config = NULL;
+    g_autofree gchar *mount_path = NULL;
+    g_autofree gchar *config_path = NULL;
+    g_autofree gchar *package_path = NULL;
+    g_autofree gchar *device_path = NULL;
+    g_autoptr(PuEmmc) emmc = NULL;
+    gchar **args;
+    gint api_version;
+    gboolean is_mounted;
+
+    if (getuid() != 0)
+        return error_not_root(error);
+
+    args = pu_command_context_get_args(context);
+    package_path = g_strdup(args[0]);
+    device_path = g_strdup(args[1]);
+
+    if (args[2]) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "Too many arguments");
+        return FALSE;
+    }
 
-    if (arg_device == NULL) {
-        arg_device = g_strdup(value);
-    } else {
-        g_set_error(error, G_OPTION_ERROR, G_OPTION_ERROR_FAILED,
-                    "%s", "Excess values in remaining arguments!");
+    if (g_strcmp0(device_path, "") <= 0) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "No device specified");
         return FALSE;
     }
 
+    if (!pu_is_drive(device_path)) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "Device '%s' is not a drive", device_path);
+        return FALSE;
+    }
+
+    if (!pu_device_mounted(device_path, &is_mounted, error)) {
+        g_prefix_error(error, "Failed checking if device is in use: ");
+        return FALSE;
+    }
+
+    if (is_mounted) {
+        g_set_error(error, PU_ERROR, PU_ERROR_FAILED,
+                    "Device '%s' is in use", device_path);
+        return FALSE;
+    }
+
+    if (!pu_package_mount(package_path, &mount_path, error))
+        return FALSE;
+
+    config_path = pu_package_get_layout_file(mount_path, error);
```

Comment:
Check if `config_path` is `NULL`.

Reply:
This should not be needed, as `pu_package_mount()` checks for an existing layout file. To make this more comprehensive and also safe from any race conditions, I'll change `pu_package_mount()` to write `config_path` directly. This also results in one less call to `pu_package_get_layout_file()`.